feat: 사용자는 환불 시에 환불액을 주고 받는다

minseok-oh opened the pull request:
## 📄 Summary

- 검증 로직
    - 요청한 사용자가 구매자가 맞는가?
    - 요청한 구매자가 거래 내역의 구매자가 맞는가?

- 구현 로직
    - 판매자는 환불액 (상품 금액 * 구매 수량)만큼 포인트를 차감한다
    - 구매자는 환불액 (상품 금액 * 구매 수량)만큼 포인트를 얻는다

## 🙋🏻 More

- 추후 구현해야 할 부분들을 TODO로 정리했습니다!
- 추후 동시성 이슈를 해결해야 하는 기능입니다!


close #60

minseok-oh committed:
feat: 환불 요청에 대한 검증 후 포인트 교환

- 환불 수량은 인자에서 제외했습니다
- 포인트 교환 이후 이루어질 task를 todo로 정리했습니다

minseok-oh committed:
refactor: 환불에 사용되는 에러 코드 정의

- 기존에 사용된 P000 에러는 입찰과 환불 모두 사용되어 거래 시로 변경합니다

minseok-oh committed:
feat: 사용자 도메인 equals 구현

- 환불 요청 사용자와 거래 내역의 사용자가 동일한 지 검증하기 위해 재정의합니다

minseok-oh committed:
test: 환불 메서드의 테스트 코드 스켈레톤 작성

HiiWee commented on src/main/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentService.java:
```diff
@@ -61,4 +55,40 @@ private boolean submitBid(long price, long auctionId, long quantity, Member buye
+
+        Member savedBuyer = memberRepository.save(buyer);
+        Member savedSeller = memberRepository.save(seller);
+        // TODO 경매 서비스에 환불 요청
+        // TODO 환불 요청에 대한 BidHistory 저장
```
제안: BidHistory에 bidStatus 필드가 있어서 status를 변경하고 저장하는 로직을 두면 환불 요청에 대한 BidHistory를 저장할 수 있을것 같아요!

HiiWee replied:
의견: 경매 서비스 환불 요청 인터페이스가 없다면 임의로 만들어도 괜찮지 않을까요?!

HiiWee commented on src/main/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentService.java:
```diff
@@ -61,4 +55,40 @@ private boolean submitBid(long price, long auctionId, long quantity, Member buye
+
+        BidHistory refundTargetBidHistory = findRefundTargetBidHistory(bidHistoryId);
+        Member refundTargetBuyer = refundTargetBidHistory.getBuyer();
+        if (buyer.equals(refundTargetBuyer)) {
+            throw new BadRequestException("환불할 입찰 내역의 구매자만 환불을 할 수 있습니다.", ErrorCode.P003);
```
의견: equals도 좋지만 BidHistory memberId만 넘겨주어 비교하는 메소드를 만들어줘도 좋을것 같아요!

HiiWee commented on src/main/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentService.java:
```diff
@@ -34,14 +34,8 @@ public void process(Member buyer, long price, long auctionId, long quantity) {
-                    .seller(savedSeller)
-                    .buyer(savedBuyer)
-                    .build();
+            BidHistory bidHistory = BidHistory.builder().productName(auctionInfo.productName()).price(price)
+                    .quantity(quantity).bidStatus(BidStatus.BID).seller(savedSeller).buyer(savedBuyer).build();
```
의견: 줄 구분이 있으면 좋을것 같아요!

HiiWee reviewed (commented):
몇개 의견 드렸읍니다 ㅎㅎ

yudonggeun commented on src/main/java/com/wootecam/luckyvickyauction/core/member/domain/Member.java:
```diff
@@ -41,4 +42,21 @@ public boolean isBuyer() {
         return this.password.equals(password);
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Member member = (Member) o;
+        return Objects.equals(id, member.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(id);
+    }
```
```java
@EqualsAndHashCode(of = "id")
```
조심스럽게 제안해봅니다. ㅎㅎ

yudonggeun reviewed (dismissed):
👍

chhs2131 commented on src/main/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentService.java:
```diff
@@ -61,4 +55,40 @@ private boolean submitBid(long price, long auctionId, long quantity, Member buye
+        long price = refundTargetBidHistory.getPrice();
+        long quantity = refundTargetBidHistory.getQuantity();
+
+        Member seller = refundTargetBidHistory.getSeller();
+        buyer.chargePoint(price * quantity);
+        seller.usePoint(price * quantity);
+
+        Member savedBuyer = memberRepository.save(buyer);
+        Member savedSeller = memberRepository.save(seller);
+        // TODO 경매 서비스에 환불 요청
+        // TODO 환불 요청에 대한 BidHistory 저장
```
좋습니당~ 포인트 수정 이후에 예외가 발생하는 경우 사용할 방어 로직도 나중에 넣어야겠네요!

minseok-oh committed:
refactor: 코드 정렬 및 경매 취소 인터페이스 정의

- todo로 넘긴 입찰 취소 요청, 정보 저장을 뼈대 코드로 구현했습니다
- 입찰 취소 요청 시 거래 내역이 경매 id를 보내줘야 하므로 필드에 추가했습니다

minseok-oh committed:
refactor: 이미 환불된 거래 내역 검증 로직 구현

- 이미 환불된 내역일 경우의 예외를 추가했습니다

minseok-oh committed:
Merge branch 'dev' into feat/60-refund-point-exchange

minseok-oh committed:
refactor: 식별자인 로그인 아이디로 동일성 검증

- signInId를 통한 동일성 검증으로 변경했습니다

minseok-oh committed:
test: 사용자 동일성 검증, 거래 내역 환불 상태 검증 테스트 추가

minseok-oh committed:
Merge remote-tracking branch 'origin/feat/60-refund-point-exchange' into feat/60-refund-point-exchange

minseok-oh committed:
Merge branch 'dev' into feat/60-refund-point-exchange
